c++/src/kj/array.h

```diff
@@ -573,6 +601,81 @@ class CappedArray {
   T content[fixedSize];
 };
 
+// =======================================================================================
+// Small-buffer-optimized SboArray and SboArrayBuilder
+//
+// SboArray and SboArrayBuilder are useful when you need a temporary buffer, whose size you
+// cannot know until runtime but is likely to be small, and whose lifetime can be bounded by either
+// the stack or some immovable parent object.
+//
+// SboArray and SboArrayBuilder are not Arrays or ArrayBuilders. In particular, they have following
+// differences:
+//
+// 1. SboArray/Builder have an inline buffer of `smallSize` elements, where `smallSize` is a size_t
+//    template parameter. If they are constructed with a size less than or equal to `smallSize`, the
+//    inline space is used, and no heap allocation is performed. Otherwise, a regular heap Array is
+//    allocated.
+//
+// 2. SboArray/Builder are immovable. You must construct them in place wherever you want to use one.
+//    They cannot be "released", "finished", or assigned-to.
+//
+// 3. SboArray/Builder have no specific constructor functions like `heapArray<T>()` or
+//    `heapArrayBuilder<T>()`. Instead, use their constructors directly, passing a single `size`
+//    parameter.
+
+
+template <typename T, size_t smallSize>
+class SboArray: private _::SboArrayDisposer<T, smallSize>,
```

Comment:
Wow... this is... quite an implementation. I would not have thought of this, but it turns out pretty concise.

I guess inheriting from SboArrayDisposer forces the SboArray to contain an extra pointer for the virtual table for this disposer. I think if you placed `T space[smallSize];` inside `SboArray` itself, then you could reference a singleton disposer, eliminating this pointer?

Reply:
Yeah, that's a better idea. I pushed 5e8076992236819efdd9aec1e6094cf619853803 with that change, plus a few more cleanup commits to simplify the allocation logic.